File: 1267/CountServersThatCommunicate.java
/*
https://leetcode.com/problems/count-servers-that-communicate/description/?envType=daily-question&envId=2025-01-23
*/

class Solution {

    public static void depthFirstSearch_forMatrices(int[][] grid, int rows, int columns, HashSet<String>seenCoordinates, int currentRow, int currentColumn, ArrayList<Integer>currentServersCount){
        for(int j=0; j<columns; ++j){
            if(grid[currentRow][j]==1){
                String tocheck=String.valueOf(currentRow)+"#"+String.valueOf(j);
                if(!seenCoordinates.contains(tocheck)){
                    currentServersCount.set(0, currentServersCount.get(0)+1);
                    seenCoordinates.add(tocheck);
                    depthFirstSearch_forMatrices(grid, rows, columns, seenCoordinates, currentRow, j, currentServersCount);
                }
            }
        }
        for(int i=0; i<rows; ++i){
            if(grid[i][currentColumn]==1){
                String tocheck=String.valueOf(i)+"#"+String.valueOf(currentColumn);
                if(!seenCoordinates.contains(tocheck)){
                    currentServersCount.set(0, currentServersCount.get(0)+1);
                    seenCoordinates.add(tocheck);
                    depthFirstSearch_forMatrices(grid, rows, columns, seenCoordinates, i, currentColumn, currentServersCount);
                }
            }
        }
    }

    public int countServers(int[][] grid) {
        int rows=grid.length, columns=grid[0].length, resultantServersCount=0;
        HashSet<String>seenCoordinates=new HashSet<>();
        for(int i=0; i<rows; ++i){
            for(int j=0; j<columns; ++j){
                if(grid[i][j]==1){
                    ArrayList<Integer>currentServersCount=new ArrayList<>();
                    currentServersCount.add(1);
                    seenCoordinates.add(String.valueOf(i)+"#"+String.valueOf(j));
                    depthFirstSearch_forMatrices(grid, rows, columns, seenCoordinates, i, j, currentServersCount);
                    if(currentServersCount.get(0)>1) resultantServersCount+=currentServersCount.get(0);
                }
            }
        }
        return resultantServersCount;
    }
}

File: 1/1000s/12s/1267/CountServersThatCommunicate.cpp
/*
https://leetcode.com/problems/count-servers-that-communicate/description/?envType=daily-question&envId=2025-01-23
*/

class Solution {
public:

    void dfs(vector<vector<int>>&grid, const int rws, const int cols, set<pair<int, int>>&seen, int currw, int curcol, int &curcount){
        for(int j=0; j<cols; ++j){
            if(grid[currw][j]==1){
                auto itr=seen.find({currw, j});
                if(itr==seen.end()){
                    curcount++;
                    seen.insert({currw, j});
                    dfs(grid, rws, cols, seen, currw, j, curcount);
                }
            }
        }
        for(int i=0; i<rws; ++i){
            if(grid[i][curcol]==1){
                auto itr=seen.find({i, curcol});
                if(itr==seen.end()){
                    curcount++;
                    seen.insert({i, curcol});
                    dfs(grid, rws, cols, seen, i, curcol, curcount);
                }
            }
        }
    }

    int countServers(vector<vector<int>>& grid) {
        int rws=grid.size(), cols=grid[0].size(), res=0;
        set<pair<int, int>>seen;
        for(int i=0; i<rws; ++i){
            for(int j=0; j<cols; ++j){
                if(grid[i][j]==1){
                    int curcount=1;
                    seen.insert({i, j});
                    dfs(grid, rws, cols, seen, i, j, curcount);
                    if(curcount>1)
                        res+=curcount;
                }
            }
        }
        return res;
    }
};